pkg/kubectl/cmd/login.go

```diff
@@ -0,0 +1,50 @@
+package cmd
+
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	utilflag "k8s.io/kubernetes/pkg/util/flag"
+)
+
+func NewCmdLogin(f *cmdutil.Factory, pathOptions *clientcmd.PathOptions, out io.Writer) *cobra.Command {
+
+	cmd := &cobra.Command{
+		Use:   "login",
+		Short: "login Log in to the cluster.",
+		Long:  `login obtains the neccesary credentials to make authenticated requests to the cluster. The type of log in depends on the 'auth-provider' type in your configuration, and will likely require some type of user interaction.`,
+
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunLogin(out, cmd, f, pathOptions, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	return cmd
+}
+
+func RunLogin(out io.Writer, cmd *cobra.Command, f *cmdutil.Factory, pathOptions *clientcmd.PathOptions, args []string) error {
+
+	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+	flags.SetNormalizeFunc(utilflag.WarnWordSepNormalizeFunc) // Warn for "_" flags
+
+	cmdCliCfg := cmdutil.DefaultClientConfig(flags)
```

Comment:
I think just the `kubectl auth login` command could make sense with some of our existing flows. For example:

If I run `kubectl auth login --auth-provider=oidc` with an empty/nonexistent kubeconfig, it could take me through an interactive flow of pointing to an apiserver and then setting up OIDC specific config. Later calls to `kubectl auth login` would use the persisted server endpoint to refresh credentials.

If I run `kubectl auth login` with no `--auth-provider` and an empty/nonexistent kubeconfig it could prompt `Please Enter Auth Provider`, and then go through the rest of the flow.

Then, it works smoothly when some outside admin program has provided you a kubeconfig like:
```
clusters:
- cluster:
    server: <some-ip>
  name: my-cluster
contexts:
- context:
    cluster: my-cluster
    user: my-user
users:
- name: my-user
  user:
      name: oidc
```
But also allows you to start from scratch/use flags.

@bobbyrullo: Does that keep to the spirit of what you want out of the `login` command?

@deads2k: Some of what you're describing sounds to me like using `kubectl auth login` as an alias for `kubectl config set users.me.{password|username}`. Am I misunderstanding?